File: broker/src/cegm_broker/_paths.py
"""Resolve filesystem paths CEGM uses at runtime.

All persistent state lives under ``%LOCALAPPDATA%\\CEGM\\`` on Windows. The
broker creates the directory tree lazily on first write. Tests can override
the root via the ``CEGM_DATA_DIR`` environment variable.
"""

from __future__ import annotations

import os
from datetime import UTC, datetime
from pathlib import Path
from typing import Final

_ENV_OVERRIDE: Final[str] = "CEGM_DATA_DIR"


def data_root() -> Path:
    """Return the base directory for all CEGM runtime state.

    On Windows this resolves to ``%LOCALAPPDATA%\\CEGM``. The directory may
    not exist yet — callers that write should use :func:`ensure_dir`.
    """
    override = os.environ.get(_ENV_OVERRIDE)
    if override:
        return Path(override)
    local_appdata = os.environ.get("LOCALAPPDATA")
    if local_appdata:
        return Path(local_appdata) / "CEGM"
    # Sensible cross-platform fallback for development on macOS/Linux.
    return Path.home() / ".local" / "share" / "CEGM"


def config_path() -> Path:
    """Return the path to the user-editable JSON config file."""
    return data_root() / "config.json"


def logs_dir() -> Path:
    """Return the directory used for daily-rotated broker logs."""
    return data_root() / "logs"


def activity_dir() -> Path:
    """Return the directory used for per-session activity event logs."""
    return data_root() / "activity"


def snapshots_dir() -> Path:
    """Return the directory used for snapshot storage."""
    return data_root() / "snapshots"


def daily_log_file(when: datetime | None = None) -> Path:
    """Return the broker log file for the given UTC date (default: now)."""
    when = when or datetime.now(UTC)
    return logs_dir() / f"broker-{when:%Y%m%d}.jsonl"


def ensure_dir(path: Path) -> Path:
    """Create ``path`` (and parents) if missing; return it for chaining."""
    path.mkdir(parents=True, exist_ok=True)
    return path

File: plugin/lib/paths.lua
--[[
  paths.lua — Resolve filesystem paths the CEGM plugin uses at runtime.

  Mirrors `cegm_broker._paths` so plugin and broker agree on where state
  lives. All persistent state sits under %LOCALAPPDATA%\CEGM\.
]]

local M = {}

---Return %LOCALAPPDATA%\CEGM (or a sensible fallback for development).
---@return string
function M.data_root()
  local override = os.getenv("CEGM_DATA_DIR")
  if override and override ~= "" then return override end
  local local_appdata = os.getenv("LOCALAPPDATA")
  if local_appdata and local_appdata ~= "" then
    return local_appdata .. "\\CEGM"
  end
  -- Fallback (non-Windows dev box).
  return (os.getenv("HOME") or ".") .. "/.local/share/CEGM"
end

---Path of the JSON config file shared with the broker.
---@return string
function M.config_path() return M.data_root() .. "\\config.json" end

---Path of the plugin-side log file (kept separate from the broker's).
---@return string
function M.plugin_log_path() return M.data_root() .. "\\logs\\plugin.log" end

---Path of the marker file the broker writes once it has bound a port.
---Plugin polls this for "is broker up?".
---@return string
function M.broker_pid_path() return M.data_root() .. "\\broker.pid" end

---Create %LOCALAPPDATA%\CEGM\ and the logs/ subdirectory if missing.
function M.ensure_data_root()
  local root = M.data_root()
  -- `mkdir` on Windows ignores existing dirs but errors with garbage if path
  -- has spaces unquoted. Use cmd /c to handle redirect of stderr safely.
  os.execute(string.format('cmd /c mkdir "%s" 2>nul', root))
  os.execute(string.format('cmd /c mkdir "%s\\logs" 2>nul', root))
end

return M
